mitiq/factories.py

```diff
@@ -162,3 +164,147 @@ def reduce(self) -> float:
+    (almost) exponential decay ansatz with a non linear exponent, i.e.:
+
+    y(x) = a + exp(z(x)), where z(x) is a polynomial of a given order.
+
+    If the asymptotic value (y(x->inf) = a) is known, a linear fit with respect
+    to z(x) := log(y(x) - a) is used.
+    Otherwise, a non-linear fit of y(x) is perfomed.
+    """
+
+    def __init__(self, scalars: Iterable[float], order: int, \
+                 asymptote: Union[float, None] = None) -> None:
+        """
+        Args:
+            scalars: Iterable of noise scale factors at which expectation values should be measured.
+            order: Polynomial extrapolation order. It cannot exceed len(scalars) - 1.
+                   If asymptote is None, order cannot exceed len(scalars) - 2.
+            asymptote: Infinite-noise limit (optional argument).
+        """
+        super(PolyDecayFactory, self).__init__(scalars)
+        if not (asymptote is None or isinstance(asymptote, float)):
+            raise ValueError("The argument 'asymptote' must be either a float or None")
+        self.order = order
+        self.asymptote = asymptote
+
+    @staticmethod
+    def static_reduce(instack: List[float], outstack: List[float], \
+                      asymptote: Union[float, None], order: int, eps: float = 1.0e-9) -> float:
+        """
+        Determines the zero-noise limit, assuming an exponential decay ansatz:
+        y(x) = a + exp(z(x)), where z(x) is a polynomial of a given order.
+        We assume that z(x->inf)=-inf, such that y(x->inf) is finite.
+
+        If self.asymptote is None, the ansatz y(x) is fitted with a non-linear optimization.
+        Otherwise, a linear fit of z(x) := log(y(x) - self.asymptote) is performed.
+
+        This static method is equivalent to the "self.reduce" method of PolyDecayFactory,
+        but can be called also by other factories which are particular cases of PolyDecayFactory,
+        e.g., DecayFactory.
+
+        Args:
+            instack: x data values.
+            outstack: y data values.
+            asymptote: y(x->inf).
+            order: extrapolation order.
+            eps: epsilon to regularize log(y - asymptote) when y <= asymptote.
+        """
+        # Shift is 0 if asymptote is given, 1 if asymptote is not given
+        shift = int(asymptote is None)
+
+        # Check arguments
+        error_str = "Data is not enough: at least two data points are necessary."
+        if instack is None or outstack is None:
+            raise ValueError(error_str)
+        if len(instack) != len(outstack) or len(instack) < 2:
+            raise ValueError(error_str)
+        if order > len(instack) - 1:
+            raise ValueError(
+                "Extrapolation order is too high. "
+                f"The order cannot exceed the number of data points minus {1 + shift}."
+            )
+
+        # CASE 1: asymptote is None.
+
+        # TODO: there must be better way of doing this. *args does not seem to work with curve_fit.
```

Comment:
I haven't tested this, but could you do a `polyfit` of the log of the data? e.g. something like
```
np.polyfit(np.log(instack), outstack, deg=order)
```
@andreamari

Reply:
This is equivalent to assume a zero asymptote and in fact one can get the fit that you proposed by setting `asymptote=0` instead of `asymptote=None`. 
If data really decays close to zero for large noise, then it should work. Otherwise I am not sure.

Reply:
Ah yes you're right. Maybe @tudorgt knows of a better way to do this.

Reply:
I think what we do is fine from a fitting point of view, it is just done in a dirty way (I hard coded 4 ansatz functions, one for each order: 0,1,2,3).

Reply:
Given that this works, I'm going to approve. If we come up with a more elegant way then it can be added to a later update